.automation/build_archi_diagram.py

```diff
@@ -0,0 +1,77 @@
+# !/usr/bin/env python3
+"""
+Generate architecture diagram of Mega-Linter
+See how to do it at https://kroki.io/#how
+"""
+import base64
+import tempfile
+import zlib
+from string import Template
+
+# pylint: disable=import-error
+import requests
+
+
+class diagram:
+    GATEWAY_SERVER = "https://kroki.io"
+    DIAGRAM_TEMPLATE = "docs/assets/archi.blockdiag.txt"
+
+    def __init__(
+        self,
+        descriptor_language_count,
+        descriptor_format_count,
+        descriptor_tooling_format_count,
+    ):
+        self.diagram_dict = dict(
+            language=str(descriptor_language_count),
+            format=str(descriptor_format_count),
+            tooling_format=str(descriptor_tooling_format_count),
+        )
+
+        self.build_diagram()
+
+    # get binary image generated following format specified
+    def get_image_bin(self, format="svg", tempfile="kroki"):
+        image_content = self.run_query(
+            self.GATEWAY_SERVER, "blockdiag", format, self.encoded_diagram
+        )
+
+        output_file = tempfile + "." + format
+        with open(output_file, "wb") as file:
+            file.write(image_content)
+
+        return output_file
+
+    # get diagram textual description
+    def get_image_text(self):
+        return self.asset
+
+    # generated image on gateway (default: online kroki.io server)
+    def run_query(self, gateway, diagram_type, diagram_format, encoded_diagram):
+        request = requests.get(
+            f"{gateway}/{diagram_type}/{diagram_format}/{encoded_diagram}"
+        )
+        if request.status_code == 200:
+            return request.content
+        else:
+            raise Exception(
+                "Query failed to run by returning code of {}".format(
+                    request.status_code
+                )
+            )
+
+    # build diagram textual description and return deflate + base64 encoded version
+    def build_diagram(self):
+        with open(self.DIAGRAM_TEMPLATE, "rt") as file:
+            diagram_source = file.read().rstrip()
+
+        self.asset = Template(diagram_source).safe_substitute(self.diagram_dict)
+
+        try:
+            temp = tempfile.NamedTemporaryFile(prefix="MegaLinter_")
+            temp.write(self.asset.encode())
+            temp.seek(0)
+            encoded_diagram = base64.urlsafe_b64encode(zlib.compress(temp.read(), 9))
```

Comment:
It may be inefficient in some cases to read back what we just wrote from the disk rather than reading it from memory. Probably best to split the base64 encode and compress calls up onto separate lines for the sake of readability. Prefer `zlib.Z_BEST_COMPRESSION` to `9` for clarity since not everyone knows their gzip compression levels by heart.

Reply:
not sure to understand all except the use of constant 👍

Reply:
I may be missing something, but I don't understand why we need to read a file that we just wrote while we still hold the contents of the file in a variable. That seems like it could be unnecessarily slow.

This is what I meant by splitting the class up onto separate lines. I know many software engineers can read everything on one line easily, but not everyone. Limiting the number of operations per line can be especially helpful for developers with certain types of disabilities, for example.

```py
compressed_diagram = zlib.compress(temp.read(), zlib.Z_BEST_COMPRESSION)
encoded_diagram = base64.urlsafe_b64encode(compressed_diagram)
```

Reply:
perharps you can help me to remove usage of temp file. If I did it, this is because I was in front of a problem that I cannot solved.
zlib.compress expect a bytes like object, and I cannot see how to pass it as argument from a variable that contains a string  Hope is it enough clear ?

Reply:
Oh, thank you for the explanation. `string.encode()` will convert a string to a UTF-8-encoded byte object, so in this case you could simply do `self.asset.encode()` as you already do above, but store the value in a variable rather than write it to a file. Does that make more sense?